src/data/bucket/fill_extrusion_bucket.js

```diff
@@ -301,7 +301,7 @@ class FillExtrusionBucket implements Bucket {
 
     addFeature(feature: BucketFeature, geometry: Array<Array<Point>>, index: number, canonical: CanonicalTileID, imagePositions: {[_: string]: ImagePosition}) {
         const flatRoof = this.enableTerrain && feature.properties && feature.properties.hasOwnProperty('type') &&
```

Comment:
Current code is a conservative hack (best effort) to enable flat roofs for some of the buildings.
Don't think it is a good idea to enable it for any feature that is larger than the building (countries, routes) - those kind of features would span across multiple tiles and we wouldn't be able to control when flat roof is used and where it is not.
> There are valid use cases to conditionally enable flat roofs using a style spec property for explicit usage, we will evaluate independently of this fix (refer #10170 (comment))

IMHO, that style spec property for explicit usage would be a great fix for this, too - I'd be concerned about regressions if merging this PR without style spec property for explicit usage.